### ISSUE TITLE
[Bug] F5TTS_v1_Base Finetune: Generated audio starts with last sentence of reference text and outputs text_to_generate out of order

### ISSUE THREAD

**Issue by DahuaYi** [SWivid/F5-TTS#1140]
### Checks

- [x] This template is only for bug reports, usage problems go with 'Help Wanted'.
- [x] I have thoroughly reviewed the project documentation but couldn't find information to solve my problem.
- [x] I have searched for existing issues, including closed ones, and couldn't find a solution.
- [x] I am using English to submit this issue to facilitate community communication.

### Environment Details

OS: Ubuntu 20.04.6 LTS  
GPU: RTX 3090  
CUDA: 11.8  
Python: 3.10  
torch==2.3.0+cu118  
F5-TTS: latest main branch  
Mixed Precision: fp16  
Training Mode: Finetune  
Model: F5TTS_v1_Base  
Tokenizer: pinyin  

### Steps to Reproduce

1)Use the F5-TTS Web UI (Finetune tab) to train a custom speaker model using your dataset.
          Training config:
                  Epochs: 200
                  Batch size per GPU: 12
                  Learning rate: 0.00001
                  Save per update: 500
                  Mixed precision: fp16
                  Model: F5TTS_v1_Base
                  Tokenizer: pinyin
2)After training completes, switch to the Inference tab.
3)Input a new sentence into the Text to Generate field.
4)Upload a reference audio for the same speaker.
5)Click Generate.
6)The generated audio starts with the last sentence from the reference text, then reads the text_to_generate, but often out of order.

### ✔️ Expected Behavior

The generated voice after training should read the content in text_to_generate from the beginning to the end, exactly as written.

### ❌ Actual Behavior

After training is complete, the generated voice always starts speaking from the last sentence of reference_text, followed by the content in text_to_generate. This results in misaligned output where text_to_generate is not cleanly and independently synthesized.

**Comment by SWivid** [comment]
> Batch size per GPU: 12

In what type, `sample`?

Can also check for many related issues/discussions, a very important aspect need to be considered is the maximum length of your dataset.

**Comment by QinhanJ** [comment]
Same happend here using a huggingface finetuned [model](https://huggingface.co/RASPIAUDIO/F5-French-MixedSpeakers-reduced/tree/main). Not all the time but 10% of time it starts with last few words of reference_text, then followed with text_to_generate.

**Comment by mikhail2013ru** [comment]
+1. I agree with the problem, I've already reported it too. The same thing happens in trained Russian fine-tune - very often the text from the reference audio file is heard at the beginning of the recording. But it happens that this happens not only at the beginning of the generated record.

**Comment by pavithirabharathi09** [comment]
Any solution for this issue?

**Comment by pavithirabharathi09** [comment]
> +1. I agree with the problem, I've already reported it too. The same thing happens in trained Russian fine-tune - very often the text from the reference audio file is heard at the beginning of the recording. But it happens that this happens not only at the beginning of the generated record.

Have you found the solution?

**Comment by SWivid** [comment]
Hi @DahuaYi , how is it going

---

The main issue may be that the attention mask is not enabled, facing issues on small datasets. This is because in a single batch, there is more padding compared to large datasets (hard to find similar length audio samples in small datasets).

---

Try use `frame` as batch_size_type
https://github.com/SWivid/F5-TTS/blob/77d3ec623b861ea3cebd3a11f0f9a14489112464/src/f5_tts/configs/F5TTS_v1_Base.yaml#L35-L36
And set `flash_attn` and `True` (install https://github.com/Dao-AILab/flash-attention first)

If you use finetune related scripts, manually (cuz finetune scripts are not using hydra yaml to configure as train.py) add the keys `attn_backend="flash_attn", attn_mask_enabled=True`, to 
https://github.com/SWivid/F5-TTS/blob/77d3ec623b861ea3cebd3a11f0f9a14489112464/src/f5_tts/train/finetune_cli.py#L91-L98

**Comment by mikhail2013ru** [comment]
> Hi [@DahuaYi](https://github.com/DahuaYi) , how is it going
> 
> The main issue may be that the attention mask is not enabled, facing issues on small datasets. This is because in a single batch, there is more padding compared to large datasets (hard to find similar length audio samples in small datasets).
> 
> Try use `frame` as batch_size_type
> 
> [F5-TTS/src/f5_tts/configs/F5TTS_v1_Base.yaml](https://github.com/SWivid/F5-TTS/blob/77d3ec623b861ea3cebd3a11f0f9a14489112464/src/f5_tts/configs/F5TTS_v1_Base.yaml#L35-L36)
> 
> Lines 35 to 36 in [77d3ec6](/SWivid/F5-TTS/commit/77d3ec623b861ea3cebd3a11f0f9a14489112464)
> 
>  attn_backend: torch  # torch | flash_attn 
>  attn_mask_enabled: False 
> 
> And set `flash_attn` and `True` (install https://github.com/Dao-AILab/flash-attention first)
> If you use finetune related scripts, manually (cuz finetune scripts are not using hydra yaml to configure as train.py) add the keys `attn_backend="flash_attn", attn_mask_enabled=True`, to
> 
> [F5-TTS/src/f5_tts/train/finetune_cli.py](https://github.com/SWivid/F5-TTS/blob/77d3ec623b861ea3cebd3a11f0f9a14489112464/src/f5_tts/train/finetune_cli.py#L91-L98)
> 
> Lines 91 to 98 in [77d3ec6](/SWivid/F5-TTS/commit/77d3ec623b861ea3cebd3a11f0f9a14489112464)
> 
>  model_cfg = dict( 
>      dim=1024 
>      depth=22, 
>      heads=16 
>      ff_mult=2, 
>      text_dim=512, 
>      conv_layers=4, 
>  )

can you add this code to the main repository? I'm running f5 tts on colab, it won't be convenient for me to edit the code there every time. Are there any plans to refine the f5 architecture so that speech sounds less mechanical? The Russians have made several finetune model. About 5 thousand hours of speech, the first and the second about 10 thousand. But it cannot be said that even 10 thousand would add to the speech of humanity. Also, whole words in sentences are often skipped in the generations.

**Comment by SWivid** [comment]
Try use train.py

**Comment by mikhail2013ru** [comment]
> Попробуйте использовать train.py

is it possible to remove the 12-second audio reference limit? And do the cloning from 12 seconds to 1 minute? It would help to copy the voice with more different intonations. Can you do this?

**Comment by SWivid** [comment]
train with 1-minute-long audio samples, current model is trained with up-to-30sec samples

**Comment by KangZENG50025543** [comment]
我已经理解问题所在了
目前这些暂时只在中文上获得了验证，英文可以参照进行修改。

在分析 F5-TTS 的外围声音-文本处理流程时，发现一个影响模型测试/使用的严重问题。


当前行为（model test 部分）：
- 当参考音频超过12秒时，会被裁剪到前12秒或更短
- 但对应的参考文本保持完整不变
- 导致模型学习到错误的映射关系：`短音频 ↔ 长文本`


这会导致：


- ❌注意力机制混乱 （生成的内容出现字段反复：Generate text:“我是守岸人，您的助理”——>Audio:“您的助理，我是守岸人，您的助理，守岸人，助理”）
- ❌时长预测失真  （生成的时长非常短，语句念得非常快）
- ❌韵律学习错误（停顿完全错误，彻底忽略标点）


问题根因：
在文件 'src/f5_tts/infer/utils_infer.py' 的 'preprocess_ref_audio_text()' 函数中：
```python
# audio cutting：
aseg = aseg[:12000]  # Or cutting based on silence

# But text remains the same：
return ref_audio, ref_text  # ❌No cutting text！
```

修复建议：
加上与之对应的文本智能裁剪，比如根据时长，标点等对应算法进行正确的裁剪。

[utils_infer_changed_version.py](https://github.com/user-attachments/files/23309762/utils_infer_changed_version.py)

**Comment by KangZENG50025543** [comment]
其实就只是单纯的reference和提供的text没有对应上。

> 使用 1 分钟长的音频样本进行训练，当前模型使用长达 30 秒的样本进行训练

**Comment by mikhail2013ru** [comment]
> На самом деле проблема заключается в том, что ссылка и предоставленный текст не соответствуют друг другу.
> 
> > Для обучения используется аудиозапись длительностью 1 минута. Текущая модель обучается на записях длительностью до 30 секунд

I suggest increasing the 1-second limit for clipping the reference file. It is unlikely that you will find a stable algorithm for working with reference text in all languages. And if the extra characters that go beyond 12 seconds remain, then this guarantees compliance in any case.

**Comment by SWivid** [comment]
> 在分析 F5-TTS 预处理流程时，发现一个影响训练质量的严重问题。

🤔 `preprocess_ref_audio_text()` is only for inference, no impact on training

**Comment by KangZENG50025543** [comment]
> > 在分析 F5-TTS 预处理流程时，发现一个影响训练质量的严重问题。
> 
> 🤔 `preprocess_ref_audio_text()` is only for inference, no impact on training

Yes. And you could find the bug happened in "Inference tab" as the DahuaYi said.

I think it's actually the bug I mentioned.

"影响训练质量" is my fault. I should say "影响模型测试/使用的严重问题"

**Comment by SWivid** [comment]
@KangZENG50025543 the original PR contributor of finetune-gradio hasn't maintained it for a long time. I have gradually fixed some fatal issues related to training (in finetune-gradio, not train.py), but there are still some problems that need to be restructured, e.g.:
- For issues related to the current one, you can directly remove the ref_text (i.e. leave blank) in the test function and call the asr to recognize it (but this will load the asr model and increase the use of device memory). So actually, users can simply switch to infer-gradio for testing. The custom model loading function is already quite complete.
- Some other points that we don't really have the energy to refactor, simply put, are to remove the finetune-cli dependency (which was actually originally a shadow of train.py), and let finetune-gradio also be managed with a unified hydra config like train.py (hopefully someone interested to do these).

**Comment by SWivid** [comment]
This issue will be tracked with 
- #1207
